### DISCUSSION TITLE
Context mutation with Nodejs streams

### DISCUSSION THREAD

**Discussion by Sonta** [statelyai/xstate#2162] [Q&A] [answered]
Hi there, I've started working with xstate quite recently and am feeling unsure if it's the right tool for the problem I'm trying to solve.

I have a Nodejs service which accepts several audio streams via websocket connections and forwards them to other services (e.g. upload to AWS S3). The main reason I am trying to implement a state machine is because websocket connections can disconnect and persisting state helps me reconnect.

I've implemented a rather simple state machine that essentially consists of a single step (I call it `play`) in which I set up the pipelines for the streams and `await` their completion. After that it just transitions into `cleanup`.

Now the issue I have come across is that I would like to change the context of the state machine, but I cannot transition out of the `play` step because I need to wait for the audio streams to finish (correct me if I'm wrong, maybe there's a trick to this which I did not know about?).

The only viable way I've found is to externally mutate the context which is an explicit anti-pattern. Is there something I have overlooked?

Best,
Son

**Comment by davidkpiano** [comment, marked as answer]
> Now the issue I have come across is that I would like to change the context of the state machine, but I cannot transition out of the `play` step because I need to wait for the audio streams to finish (correct me if I'm wrong, maybe there's a trick to this which I did not know about?).

Do you have some code to share for this? You can assign to context without changing states:

```js
on: {
  SOME_EVENT: {
    actions: assign({ /* change context here */ })
  }
}
```

Would love to help with your use-case. Especially with WebSockets, managing the connection states of them makes good use of an explicit state machine.

**Reply by Sonta** [reply]
Hi, thanks for the quick reply!

Correct me if I'm wrong, but assigning to context does not mutate it but creates a new context right? This would therefore not influence any already-running processes, right?

What I've also observed is that the state machine does self-transition after an `assign`, which is what I expected according to https://xstate.js.org/docs/guides/transitions.html#internal-transitions. I'm logging all events and after `MY_EVENT` is received by the machine I do see a state transition from `play` to `play` which makes me lose access to the already running processes.

Here's the (I believe) relevant xstate code snippet:

```
const machine = Machine<SessionMachineContext, SessionMachineStateSchema, SessionMachineEvent>({
  context: {
    sessionContext,
    error: undefined,
  },
  initial: 'new' as const,
  states: {
    new: {
      always: 'play',
    },
    play: {
      invoke: {
        id: 'play',
        src: (context) => play(context, session),
      },
      on: {
        MY_EVENT: {
          actions: assign({
            sessionContext: (context, event) => {
              context.sessionContext.data = event.data;
              return context.sessionContext;
            },
          }),
        },
      },
    },
  },
})
```

**Reply by Andarist** [reply]
Such self-transition doesnt exit its source state so you shouldnt lose access to anything there. If you do - please create a runnable repro case that we could investigate.

**Reply by Sonta** [reply]
Oh wow, I really assumed the worst before trying it out, shame on me!

Tried it out and self-transition does indeed preserve access to the closure, quite fascinating to me actually. And the reference passing also works like a charm, so even though the `context` was re-assigned the consuming child process is updated automatically, so everything works :)

Thanks a lot, this discussion can be closed.
